django/db/models/query.py

```diff
@@ -995,9 +995,10 @@ def update_or_create(self, defaults=None, create_defaults=None, **kwargs):
                 # This is to maintain backward compatibility as these fields
                 # are not updated unless explicitly specified in the
                 # update_fields list.
+                pk_fields = self.model._meta.pk_fields
                 for field in self.model._meta.local_concrete_fields:
                     if not (
-                        field.primary_key or field.__class__.pre_save is Field.pre_save
+                        field in pk_fields or field.__class__.pre_save is Field.pre_save
```

Comment:
This one is right because we iterate over `local_concrete_fields` thus inherited `primary_key` is not included.

If we were to switch to `concrete_fields` as proposed in ticket-35890 we'd have to make adjustments though.